Tools/ssl/make_ssl_data.py

```diff
@@ -55,103 +109,178 @@ def _file_search(fname, pat):
-        if "_F_" in reason:
-            # ignore function codes
+        if "_F_" in reason:  # ignore function codes
+            # FEAT(picnixz): in the future, we may want to also check
+            # the consistency of the OpenSSL files with an external tool.
             continue
-        num = int(num)
-        yield reason, libname, errname, num
+        yield reason, libname, errname, int(num)
 
 
 def parse_extra_reasons(args):
-    """Parse extra reasons from openssl.ec"""
+    """Parse extra reasons from crypto/err/openssl.ec.
+
+    Detected lines are matched against "R <LIBNAME>_R_<ERRNAME> <ERRCODE>",
+    e.g., "R SSL_R_SSLV3_ALERT_UNEXPECTED_MESSAGE 1010".
+    """
     pat = re.compile(r"^R\s+((\w+)_R_(\w+))\s+(\d+)")
-    for match in _file_search(args.errcodes, pat):
+    for match in file_search(args.errcodes, pat):
         reason, libname, errname, num = match.groups()
-        num = int(num)
-        yield reason, libname, errname, num
+        yield reason, libname, errname, int(num)
 
 
 def gen_library_codes(args):
-    """Generate table short libname to numeric code"""
+    """Generate table short libname to numeric code."""
+    yield "/* generated from args.lib2errnum */"
     yield "static struct py_ssl_library_code library_codes[] = {"
     for libname in sorted(args.lib2errnum):
         yield f"#ifdef ERR_LIB_{libname}"
         yield f'    {{"{libname}", ERR_LIB_{libname}}},'
         yield "#endif"
-    yield "    { NULL }"
+    yield "    {NULL, 0}  /* sentinel */"
     yield "};"
-    yield ""
 
 
 def gen_error_codes(args):
-    """Generate error code table for error reasons"""
+    """Generate error code table for error reasons."""
+    yield "/* generated from args.reasons */"
     yield "static struct py_ssl_error_code error_codes[] = {"
     for reason, libname, errname, num in args.reasons:
-        yield f"  #ifdef {reason}"
+        yield f"#ifdef {reason}"
         yield f'    {{"{errname}", ERR_LIB_{libname}, {reason}}},'
-        yield "  #else"
+        yield "#else"
         yield f'    {{"{errname}", {args.lib2errnum[libname]}, {num}}},'
-        yield "  #endif"
-
-    yield "    { NULL }"
+        yield "#endif"
+    yield "    {NULL, 0, 0}  /* sentinel */"
     yield "};"
-    yield ""
 
 
-def main():
-    args = parser.parse_args()
-
+def get_openssl_git_commit(args):
+    git_describe = subprocess.run(
+        ['git', 'describe', '--long', '--dirty'],
+        cwd=args.srcdir,
+        capture_output=True,
+        encoding='utf-8',
+        check=True,
+    )
+    # FEAT(picnixz): in the future, we may want to check that we only
+    # generate data from a Git commit that is a tag (to that end, we
+    # can either re-run 'git describe --long --dirty --contains' and
+    # compare the output (the HEAD case is special as it is never
+    # contained) or parse the output of 'git describe' to count
+    # the number of commits until the previous tag.
```

Comment:
Why block running the script on an in-development version of OpenSSL? We should only *release* from a tag, of course, but I don't think this script should enforce that.

Reply:
Some mnemonics may change in an in-development version (and may even be renumbered). But I can print a warning instead of blocking.

Reply:
NVM, I never committed the bits that were responsible for raising :') I can remove the comment though and track this in an issue for discussion.